backends-velox/src/main/scala/org/apache/gluten/datasource/ArrowConvertorRule.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gluten.datasource
+
+import org.apache.gluten.backendsapi.BackendsApiManager
+import org.apache.gluten.sql.shims.SparkShimLoader
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.sql.SparkSession
+import org.apache.spark.sql.catalyst.csv.CSVOptions
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.util.PermissiveMode
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, LogicalRelation}
+import org.apache.spark.sql.execution.datasources.csv.CSVFileFormat
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.utils.SparkSchemaUtil
+
+import java.nio.charset.StandardCharsets
+
+@Experimental
+case class ArrowConvertorRule(session: SparkSession) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+    if (!BackendsApiManager.getSettings.enableNativeArrowReadFiles()) {
+      return plan
+    }
+    plan.resolveOperators {
+      // Read path
+      case l @ LogicalRelation(
+            r @ HadoopFsRelation(_, _, dataSchema, _, _: CSVFileFormat, options),
+            _,
+            _,
+            _) =>
+        val csvOptions = new CSVOptions(
+          options,
+          columnPruning = session.sessionState.conf.csvColumnPruning,
+          session.sessionState.conf.sessionLocalTimeZone)
+        if (
+          checkSchema(dataSchema) &&
+          checkCsvOptions(csvOptions, session.sessionState.conf.sessionLocalTimeZone)
+        ) {
+          l.copy(relation = r.copy(fileFormat = new ArrowCSVFileFormat())(session))
+        } else l
+      case r =>
+        r
+    }
+  }
+
+  private def checkCsvOptions(csvOptions: CSVOptions, timeZone: String): Boolean = {
+    csvOptions.headerFlag && !csvOptions.multiLine && csvOptions.delimiter == "," &&
```

Comment:
Does Arrow only support "," delimiter?

Reply:
Yes, this is the default delimiter

Reply:
And I will discuss with Arrow community to support more options, https://github.com/apache/arrow/issues/28866#issuecomment-2095354665